include/CLI/App.hpp

```diff
@@ -1720,44 +1778,58 @@ class App {
             app->_configure();
         }
     }
-    /// Internal function to run (App) callback, top down
+    /// Internal function to run (App) callback, bottom up
     void run_callback() {
         pre_callback();
-        if(callback_)
-            callback_();
+        // run the callbacks for the received subcommands
         for(App *subc : get_subcommands()) {
-            subc->run_callback();
+            if(!subc->immediate_callback_)
+                subc->run_callback();
+        }
+        // now run callbacks for option_groups
+        for(auto &subc : subcommands_) {
+            if((!subc->immediate_callback_) && (subc->name_.empty()) && (subc->count_all() > 0)) {
+                subc->run_callback();
+            }
+        }
+        // finally run the main callback
+        if(callback_ && (parsed_ > 0)) {
+            if((!name_.empty()) || (count_all() > 0)) {
+                callback_();
+            }
         }
     }
 
     /// Check to see if a subcommand is valid. Give up immediately if subcommand max has been reached.
-    bool _valid_subcommand(const std::string &current) const {
+    bool _valid_subcommand(const std::string &current, bool ignore_used = true) const {
         // Don't match if max has been reached - but still check parents
         if(require_subcommand_max_ != 0 && parsed_subcommands_.size() >= require_subcommand_max_) {
-            return parent_ != nullptr && parent_->_valid_subcommand(current);
+            return parent_ != nullptr && parent_->_valid_subcommand(current, ignore_used);
         }
-        auto com = _find_subcommand(current, true, true);
+        auto com = _find_subcommand(current, true, ignore_used);
         if(com != nullptr) {
             return true;
         }
         // Check parent if exists, else return false
-        return parent_ != nullptr && parent_->_valid_subcommand(current);
+        return parent_ != nullptr && parent_->_valid_subcommand(current, ignore_used);
     }
 
     /// Selects a Classifier enum based on the type of the current argument
-    detail::Classifier _recognize(const std::string &current) const {
+    detail::Classifier _recognize(const std::string &current, bool ignore_used_subcommands = true) const {
         std::string dummy1, dummy2;
 
         if(current == "--")
             return detail::Classifier::POSITIONAL_MARK;
-        if(_valid_subcommand(current))
+        if(_valid_subcommand(current, ignore_used_subcommands))
             return detail::Classifier::SUBCOMMAND;
         if(detail::split_long(current, dummy1, dummy2))
             return detail::Classifier::LONG;
         if(detail::split_short(current, dummy1, dummy2))
             return detail::Classifier::SHORT;
         if((allow_windows_style_options_) && (detail::split_windows_style(current, dummy1, dummy2)))
             return detail::Classifier::WINDOWS;
+        if((current == "++") && (!name_.empty()))
```

Comment:
```suggestion
        if((current == "++") && !name_.empty())
```

(Make it slightly easier to parse by eye)